pkg/sign/transparency.go

```diff
@@ -0,0 +1,198 @@
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/pki"
+	"github.com/sigstore/rekor/pkg/types"
+	"github.com/sigstore/rekor/pkg/types/dsse"
+	"github.com/sigstore/rekor/pkg/types/hashedrekord"
+	"github.com/sigstore/rekor/pkg/util"
+
+	// To initialize rekor types
+	_ "github.com/sigstore/rekor/pkg/types/dsse/v0.0.1"
+	_ "github.com/sigstore/rekor/pkg/types/hashedrekord/v0.0.1"
+)
+
+type Transparency interface {
+	GetTransparencyLog([]byte, *protobundle.Bundle) error
+}
+
+type Rekor struct {
+	options *RekorOptions
+}
+
+type RekorOptions struct {
+	// URL of Fulcio instance
+	BaseURL string
+	// Optional timeout for network requests
+	Timeout time.Duration
+	// Optional version string for user agent
+	LibraryVersion string
+}
+
+func NewRekor(opts *RekorOptions) *Rekor {
+	return &Rekor{options: opts}
+}
+
+func (r *Rekor) GetTransparencyLog(pubKeyPEM []byte, b *protobundle.Bundle) error {
+	artifactProperties := types.ArtifactProperties{
+		PublicKeyBytes: [][]byte{pubKeyPEM},
+	}
+
+	dsseEnvelope := b.GetDsseEnvelope()
+	messageSignature := b.GetMessageSignature()
+	verificationMaterial := b.GetVerificationMaterial()
+	bundleCertificate := verificationMaterial.GetCertificate()
+
+	var proposedEntry models.ProposedEntry
+	var kind, version string
+
+	switch {
+	case dsseEnvelope != nil:
+		dsseType := dsse.New()
+		kind = dsse.KIND
+		version = dsseType.DefaultVersion()
+
+		artifactBytes, err := json.Marshal(dsseEnvelope)
+		if err != nil {
+			return err
+		}
+
+		artifactProperties.ArtifactBytes = artifactBytes
+
+		proposedEntry, err = dsseType.CreateProposedEntry(context.TODO(), "", artifactProperties)
+		if err != nil {
+			return err
+		}
+	case messageSignature != nil:
+		hashedrekordType := hashedrekord.New()
+		kind = hashedrekord.KIND
+		version = hashedrekordType.DefaultVersion()
+
+		if bundleCertificate == nil {
+			return errors.New("hashedrekord requires X.509 certificate")
+		}
+
+		hexDigest := hex.EncodeToString(messageSignature.MessageDigest.Digest)
+
+		artifactProperties.PKIFormat = string(pki.X509)
+		artifactProperties.SignatureBytes = messageSignature.Signature
+		artifactProperties.ArtifactHash = util.PrefixSHA(hexDigest)
+
+		var err error
+		proposedEntry, err = hashedrekordType.CreateProposedEntry(context.TODO(), "", artifactProperties)
+		if err != nil {
+			return err
+		}
+	default:
+		return errors.New("unable to find signature in bundle")
+	}
+
+	params := entries.NewCreateLogEntryParams()
+	if r.options.Timeout > 0 {
+		params.SetTimeout(r.options.Timeout)
+	}
+	params.SetProposedEntry(proposedEntry)
+
+	client, err := client.GetRekorClient(r.options.BaseURL, client.WithUserAgent(constructUserAgent(r.options.LibraryVersion)))
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Entries.CreateLogEntry(params)
+	if err != nil {
+		return err
+	}
+
+	entry := resp.Payload[resp.ETag]
+	keyID, err := hex.DecodeString(*entry.LogID)
+	if err != nil {
+		return err
+	}
+
```

Comment:
Can you use https://github.com/sigstore/rekor/blob/92cf4d5d3682ffcbf4138341f1bc5f7a66766549/pkg/tle/tle.go#L33 instead to construct the TLE struct?